### PR TITLE
Some precedence refactoring+fixes

### PR DESCRIPTION
**Pull request by justinj**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3087)

### CODE REVIEW

File: src/sql-parser/tests/sqlparser_common.rs [hunk 1]
```diff
@@ -104,6 +105,37 @@ Expected INTO, found: public"
...
+        ("a + b * c", "a + (b * c)"),
+        ("true = 'foo' like 'foo'", "true = ('foo' like 'foo')"),
+        ("a->b = c->d", "(a->b) = (c->d)"),
+        ("a @> b = c @> d", "(a @> b) = (c @> d)"),
+        ("a = b is null", "(a = b) is null"),
+        ("a and b or c and d", "(a and b) or (c and d)"),
+    ] {
+        let left = Parser::parse_sql(format!("SELECT {}", actual)).unwrap();
+        let mut right = Parser::parse_sql(format!("SELECT {}", expected)).unwrap();
+        VisitMut::visit_statement(&mut RemoveParens, &mut right[0]);
```
Comment (benesch) [inline src/sql-parser/tests/sqlparser_common.rs:133]: `RemoveParens.visit_statement(&mut right[0])` is "idiomatic"

File: src/sql-parser/tests/sqlparser_common.rs [hunk 2]
```diff
@@ -104,6 +105,37 @@ Expected INTO, found: public"
...
+
+    for (actual, expected) in &[
+        ("a + b + c", "(a + b) + c"),
+        ("a - b + c", "(a - b) + c"),
+        ("a + b * c", "a + (b * c)"),
+        ("true = 'foo' like 'foo'", "true = ('foo' like 'foo')"),
+        ("a->b = c->d", "(a->b) = (c->d)"),
+        ("a @> b = c @> d", "(a @> b) = (c @> d)"),
+        ("a = b is null", "(a = b) is null"),
+        ("a and b or c and d", "(a and b) or (c and d)"),
```
Comment (benesch) [inline src/sql-parser/tests/sqlparser_common.rs:129]: Ha! This is clever. There's a bunch of uglier precedence tests scattered throughout this file that perhaps you'd like to absorb here?

File: src/sql-parser/src/parser.rs [hunk 3]
```diff
@@ -101,6 +101,21 @@ pub struct Parser {
     index: usize,
 }
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Precedence {
+    Zero,
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:106]: Perhaps throw an "order controls the precedence" comment in here? It's easy to read past the `Ord` derivation.

File: src/sql-parser/src/parser.rs [hunk 4]
```diff
@@ -834,35 +836,33 @@ impl Parser {
...
+                | Token::JsonContainsPath => Precedence::JsonCompare,
+                Token::JsonGet
+                | Token::JsonGetAsText
+                | Token::JsonGetPath
+                | Token::JsonGetPathAsText
                 | Token::JsonConcat
-                | Token::JsonDeletePath
-                | Token::JsonContainsPath => Precedence::Json,
-
+                | Token::JsonDeletePath => Precedence::JsonOp,
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:862]: Unrelated, and you would know better than I, but Postgres probably has names for these ops that are more general than just JSON ops, right?

**Comment by justinj** [comment]
Yeah, I think "contains" is a good word for it, I think it's also used for geometry types?

**Pull request merged**